src/smc-hub/kucalc/quota.ts

```diff
@@ -85,10 +86,94 @@ interface Settings {
   cpu_shares?: string;
 }
 
+interface Upgrades {
+  disk_quota: number;
+  memory: number;
+  memory_request: number;
+  cores: number;
+  network: number;
+  cpu_shares: number;
+  mintime: number;
+  member_host: number;
+  ephemeral_state: number;
+  ephemeral_disk: number;
+}
+
+// special quotas for on-prem setups.
+// They not only set different defaults,
+// but also tune some aspects of the overall behavior.
+interface SiteSettingsDefaultQuotas {
+  internet: boolean; // true, allow project pods to access the internet
+  idle_timeout: number; // overrides DEFAULT_QUOTAS.mintime
+  cpu: number; // shared cpu quota, in 1 core units, overrides DEFAULT_QUOTAS.cores
+  cpu_oc: number; // overcommitment ratio for cpu, 1:cpu_oc
+  mem: number; // shared memory quota, in mb, overrides DEFAULT_QUOTAS.memory
+  mem_oc: number; // overcommitment ratio for memory, 1:mem_oc
+  disk_quota: number; // overrides DEFAULT_QUOTAS.disk_quota
+}
+
+/*
+ * default quotas: {"internet":true,"mintime":3600,"mem":1000,"cpu":1,"cpu_oc":10,"mem_oc":5}
+ * max_upgrades: Quota
+ */
+interface SiteSettingsQuotas {
+  default_quotas: Partial<SiteSettingsDefaultQuotas>;
+  max_upgrades: Partial<Upgrades>;
+}
+
+// base quota + calculated default quotas is the quota object each project gets by default
+// any additional quotas are added on top of it, up until the given limits
+const BASE_QUOTAS: Quota = {
+  network: false,
+  member_host: false,
+  memory_request: 0, // will hold guaranteed RAM in MB
+  cpu_request: 0, // will hold guaranteed min number of vCPU's as a float from 0 to infinity.
+  privileged: false, // for elevated docker privileges (FUSE mounting, later more)
+  disk_quota: DEFAULT_QUOTAS.disk_quota,
+  memory_limit: DEFAULT_QUOTAS.memory, // upper bound on RAM in MB
+  cpu_limit: DEFAULT_QUOTAS.cores, // upper bound on vCPU's
+  idle_timeout: DEFAULT_QUOTAS.mintime // minimum uptime
+} as const;
+
+function calc_default_quotas(site_settings?: SiteSettingsQuotas): Quota {
+  const q: Quota = Object.assign({}, BASE_QUOTAS);
+
+  // overwrite/set extras for any set default quota in the site setting
+  if (site_settings != null && site_settings?.default_quotas != null) {
+    const dq = site_settings?.default_quotas;
```

Comment:
Aren't you in an if statement where you already checked that site_settings is defined?

Reply:
(If you're working around shortcomings in Typescript, I would say so explicitly in a comment...)